src/file_readers/spectrum_readers/dda_reader/precursors.rs

```diff
@@ -29,6 +29,7 @@ impl PrecursorReader {
         let precursor_table: PrecursorTable =
             PrecursorTable::from_sql(&tdf_reader.tdf_sql_reader);
         let retention_times: Vec<f64> = tdf_reader.frame_table.rt.clone();
+        let collision_energy = &pasef_frames.collision_energy_by_precursor;
```

Comment:
This is probably yhe place to do `group by precursor` you currently do in the table loading.